doc/specifications/proposed/named_queries/developer_documentation.md

```diff
@@ -0,0 +1,180 @@
+- `QueryType\Parameter`: matches the value of one or several of the parameters passed to the QueryType
+- `QueryType\Expression`: matches the parameters with an [Expression](http://symfony.com/en/doc/current/components/expression_language/index.html).
+
+Assuming we want to render the `AcmeBundle:LatestContent` query using
+`AcmeBundle:query/list/latest_articles.html.twig` when the `ContentType` parameter has the value 'article':
+
+```twig
+system:
+    default:
+        query:
+            list:
+                latest_articles:
+                    template: "AcmeBundle:query/list/latest_articles.html.twig"
+                    match:
+                        QueryType\Name: 'AcmeBundle:LatestContent'
+                        QueryType\Parameters: {ContentType: "article"}
+```
+
+#### Template matchers
+
+##### `QueryType\Name`
+Matches the QueryType's name (exact match).
+
+##### `QueryType\Parameters`
+Matches the parameters against a given hash. The hash must contain the parameter's name as the key,
+and the match value as the value:
+
+```twig
+match:
+  QueryType\Parameters:
+    type: "article",
+    category: "development"
+
+Each parameter will be matched exactly.
+The matcher will match if ALL of the provided parameters match what is contained in the hash.
+
+##### `QueryType\Expression`
+Uses Symfony's [Expression Language](http://symfony.com/en/doc/current/components/expression_language/index.html)
+for advanced matching.
+This matcher expects a valid expression language string as the input. The expression must be
+evaluated to a boolean. The parameters hash is available as `parameters`
+
+In the example below, 'type' must be either 'article' or 'blog_post'.
+
+```twig
+match:
+  QueryType\Expression: "parameters['type'] in ['article', 'blog_post']"
+```
+
+### Query controller templates
+The templates used by this controller are mostly meant as container for more sub-requests. Except for very simple cases
+(bullet points, table), you should use a sub-request to the Content controller on the query results.
+
+#### Location Query
+The query results are available as `location_array`, an array of `Location` objects.
+
+#### Content Query
+The Query results are available as `content_array`, an array of ContentInfo objects.
+
+## QueryType objects
+To make a new QueryType available to the Query Controller, you need to register a new
+QueryType and register it.
+
+A QueryType is a PHP class that implements QueryType interface (`eZ\Publish\Core\MVC\Query\QueryType`).
+It has two methods:
+- `getName(): string`
+  Returns the QueryType name.
+- `getQuery( array $parameters = null ): Query`
+  It accepts an optional hash of parameters that can be used in any way in the implementation:
+  - customize an element's value (limit, ContentType identifier, etc)
+  - conditionally add/remove criteria from the query
+  - set the limit/offset...
+  It returns a Query object, that can be either a LocationQuery or a (Content) Query.
```

Comment:
What would that look like ? 

```php
$options = $queryType->configureOptions( [...] );
$query = $queryType->getQuery( $options );
```

I kind of expected that the OptionsResolver *might* be used internally by QueryTypes, but it would be transparent for the user.